### DISCUSSION TITLE
Protocol with callable attribute that has a recursive return type

### DISCUSSION THREAD

**Discussion by tjsmart** [microsoft/pyright#6126] [Q&A]
I have a use case for creating a protocol which has an attribute (`func` in my example below) that will return another instance of this same protocol. Below is a script where I have two attempts at defining such a protocol `CallProto` and `MethProto` along with two possible implementations that I would like to pass the type check `Impl1` and `Impl2`.

I'm running into some errors that I don't think I quite understand and could use some help. Specifically,
1. what do these pyright errors mean?
2. how can I write a protocol that accepts `Impl1` and `Impl2`?
3. when should I use Callable vs. defining a method in a protocol definition?

Links to documentation are welcome 🙂.

Here is my script (pyright errors are given below and inlayed as comments):
```python
from __future__ import annotations
from collections.abc import Callable
from typing import Protocol

class CallProto(Protocol):
    func: Callable[[], CallProto]

class MethProto(Protocol):
    def func(self) -> MethProto:
        ...

class Impl1:
    def func(self) -> Impl1:
        return Impl1()

class Impl2:
    class CallableClass:
        def __call__(self) -> Impl2:
            return Impl2()

    func = CallableClass()


def requires_call_proto(a: CallProto) -> None:
    ...

def requires_meth_proto(b: MethProto) -> None:
    ...

requires_call_proto(Impl1())  # Argument of type "Impl1" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
requires_meth_proto(Impl1())

requires_call_proto(Impl2())  # Argument of type "Impl2" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
requires_meth_proto(Impl2())  # Argument of type "Impl2" cannot be assigned to parameter "b" of type "MethProto" in function "requires_meth_proto"
```

Here is the full pyright output:
```console
$ pyright t.py
/tmp/pyright/t.py
  /tmp/pyright/t.py:30:21 - error: Argument of type "Impl1" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
    "Impl1" is incompatible with protocol "CallProto"
      "func" is invariant because it is mutable
      "func" is an incompatible type
        Type "() -> Impl1" cannot be assigned to type "() -> CallProto"
          Function return type "Impl1" is incompatible with type "CallProto"
            "Impl1" is incompatible with protocol "CallProto" (reportGeneralTypeIssues)
  /tmp/pyright/t.py:33:21 - error: Argument of type "Impl2" cannot be assigned to parameter "a" of type "CallProto" in function "requires_call_proto"
    "Impl2" is incompatible with protocol "CallProto"
      "func" is invariant because it is mutable
      "func" is an incompatible type
        Type "CallableClass" cannot be assigned to type "() -> CallProto"
          Function return type "Impl2" is incompatible with type "CallProto"
            "Impl2" is incompatible with protocol "CallProto" (reportGeneralTypeIssues)
  /tmp/pyright/t.py:34:21 - error: Argument of type "Impl2" cannot be assigned to parameter "b" of type "MethProto" in function "requires_meth_proto"
    "Impl2" is incompatible with protocol "MethProto"
      "func" is an incompatible type
        Type "CallableClass" cannot be assigned to type "(self: Impl2) -> MethProto"
          Function accepts too many positional parameters; expected 0 but received 1 (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations
```

Other details:
```console
$ lsb_release --description
Description:    Ubuntu 20.04.6 LTS

$ python --version
Python 3.10.13

$ pyright --version
pyright 1.1.316
```

While I prefer to use `pyright` I also checked the output of `mypy` out of curiousity:
```console
$ mypy t.py
t.py:30: error: Argument 1 to "requires_call_proto" has incompatible type "Impl1"; expected "CallProto"  [arg-type]
t.py:30: note: Following member(s) of "Impl1" have conflicts:
t.py:30: note:     Expected:
t.py:30: note:         def () -> CallProto
t.py:30: note:     Got:
t.py:30: note:         def func(self) -> Impl1
t.py:30: note: Protocol member CallProto.func expected settable variable, got read-only attribute
t.py:33: error: Argument 1 to "requires_call_proto" has incompatible type "Impl2"; expected "CallProto"  [arg-type]
t.py:33: note: Following member(s) of "Impl2" have conflicts:
t.py:33: note:     func: expected "Callable[[], CallProto]", got "CallableClass"
Found 2 errors in 1 file (checked 1 source file)
```

In this case `mypy` accepts both `Impl1` and `Impl2` as implementations of `MethProto` which is what I'm hoping for but I don't know if that is correct. 🤷‍♂️

**Comment by erictraut** [comment]
Mypy is correct here. `Impl1` and `Impl2` should both be considered incompatible with `CallProto`, but both of them should be compatible with `MethodProto`. There is a bug in pyright's protocol matching that affects the case where the protocol uses a `def` statement but the implementation uses a callback protocol (or an object with a `__call__` method). 

I'm converting this to [a bug report](https://github.com/microsoft/pyright/issues/6127) so we can track it accordingly.

**Reply by erictraut** [reply]
This will be fixed in the next release of pyright.

**Reply by tjsmart** [reply]
Awesome, thank you for the incredibly fast response! So it looks like the `MethodProto` is the right way to go here. But I guess I am still curious about what is the difference between defining a protocol with a method `func` or an attribute `func` that is a `Callable`.

For example, if I simply make the return type a str:
```python
from __future__ import annotations
from collections.abc import Callable
from typing import Protocol

R = str

class CallProto(Protocol):
    func: Callable[[], R]

class MethProto(Protocol):
    def func(self) -> R:
        ...

class Impl1:
    def func(self) -> R:
        return "hello"

class Impl2:
    class CallableClass:
        def __call__(self) -> R:
            return "goodbye"

    func = CallableClass()

def requires_call_proto(a: CallProto) -> None:
    ...

def requires_meth_proto(b: MethProto) -> None:
    ...

requires_call_proto(Impl1())
requires_meth_proto(Impl1())

requires_call_proto(Impl2())
requires_meth_proto(Impl2())
```
I get the output:
```console
$ pyright y.py
/tmp/pyright/y.py
  /tmp/pyright/y.py:35:21 - error: Argument of type "Impl2" cannot be assigned to parameter "b" of type "MethProto" in function "requires_meth_proto"
    "Impl2" is incompatible with protocol "MethProto"
      "func" is an incompatible type
        Type "CallableClass" cannot be assigned to type "(self: Impl2) -> R"
          Function accepts too many positional parameters; expected 0 but received 1 (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Now using `CallProto` is perfectly fine but then `MethProto` fails on `Impl2`. 🤷

Based on pyright's error message it seems that `Callable` is invariant in what the return type is, so returning something that is a subtype is denied. And I assume that the method allows for variance in the return type. Is this expected behavior something documented somewhere? My googling is leading me nowhere.

**Reply by erictraut** [reply]
This error has to do with the vagaries of callables and class-scoped variables in Python. In most languages, you need to specify whether a class-scoped variable is a class var or an instance var. In Python, it's much more complicated because you can have a class var that is "shadowed" by an instance var.

```python
class Foo:
    x: int = 3

f = Foo()
print(Foo.x)  # 3
print(f.x)  # 3

Foo.x = 2
print(Foo.x)  # 2
print(f.x)  # 2

f.x = 1
print(Foo.x)  # 2
print(f.x)  # 1

del f.x
print(Foo.x)  # 2
print(f.x)  # 2
```

In the above example, the class var and the instance var are all type `int`, and regardless of whether you access the value from the class `Foo` or the instance `f`, you always get an `int` value.

Now consider a new example where we change the type of `x` to `Callable[[], None]`. 

```python
from typing import Callable

class Foo:
    x: Callable[[], int] = lambda: 3

f = Foo()
print(Foo.x())  # 3

try:
    print(f.x())  # Crash!
except:
    pass

f.x = lambda: 1
print(f.x())  # 1
```

The `Callable` special form doesn't provide a mechanism to specify whether the callable should act like a static method, a class method, or an instance method. Type checkers need to decide how to interpret `Callable` based on context. PEP 544 (which introduced `Protocol` classes) indicates that a variable within a protocol definition should be interpreted as an instance variable unless it is explicitly declared using the `ClassVar` special form.

This is important because the runtime treats a callable differently depending on whether it's accessed through a class or an instance, whether it's a static method, class method, or instance method, and whether it's stored as a class variable or an instance variable.

I'm not sure if that was a clear explanation. It's quite complicated and difficult to explain. You may find it useful to read [this section of the pyright documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=class-and-instance-variables).

Here's another code sample that might help make this clearer.

```python
from typing import Callable

class Foo:
    def method(self) -> None:
        ...

    # Both mypy and pyright consider this a type violation because `method`
    # accepts a single argument `self`.
    func: Callable[[], None] = method  # Incompatible type in assignment
```

**Reply by tjsmart** [reply]
Interesting, I think I better understand the different considerations of class vs. instance methods/callables for a protocol.

However, I am still struggling to connect the dots as to why using a sub-type is allowed in one case versus the other.

I tried taking another crack at simplifying the example even more. Here's the example using a `Callable` in my protocol:
```python
from __future__ import annotations

from collections.abc import Callable
from typing import Protocol

class CallProtoSimple(Protocol):
    func: Callable[[], str]

class MyStr(str):
    ...

class ImplStr:
    def func(self) -> str:
        return "test"

class ImplMyStr:
    def func(self) -> MyStr:
        return MyStr("test")

def requires_call_proto(a: CallProtoSimple) -> None:
    ...

requires_call_proto(ImplStr())
requires_call_proto(ImplMyStr())
```

`pyright` gives the following error regarding the use of `MyStr` instead of `str`:
```console
$ pyright a.py
/tmp/pyright/a.py
  /tmp/pyright/a.py:24:21 - error: Argument of type "ImplMyStr" cannot be assigned to parameter "a" of type "CallProtoSimple" in function "requires_call_proto"
    "ImplMyStr" is incompatible with protocol "CallProtoSimple"
      "func" is invariant because it is mutable
      "func" is an incompatible type
        Type "() -> MyStr" cannot be assigned to type "() -> str"
          Function return type "MyStr" is incompatible with type "str"
            "MyStr" is incompatible with "str" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Then I switch to using `def`'s in the protocol definition and then everything is "okay":
```python
from __future__ import annotations

from typing import Protocol

class MethProtoSimple(Protocol):
    def func(self) -> str:
        ...

class MyStr(str):
    ...

class ImplStr:
    def func(self) -> str:
        return "test"

class ImplMyStr:
    def func(self) -> MyStr:
        return MyStr("test")

def requires_meth_proto(a: MethProtoSimple) -> None:
    ...

requires_meth_proto(ImplStr())
requires_meth_proto(ImplMyStr())
```
```console
$ pyright b.py
0 errors, 0 warnings, 0 informations
```

Why does `MethProtoSimple` allow for the subclass `MyStr` but `CallableProtoSimple` does not?

I feel like I must be missing something obvious 😅 . No worries if you think the answer has already been explained, I can think about it more.

Really appreciate the explanation and bugfix!
